puppet: Add documentation and remove deprecation for show_diff, keep deprecation for alias show-diff

vaudelbenoit opened the pull request:
First, I'm new to this contribution thing and although I'm trying to follow the PR quick start guide I wouldn't be surprised if I'm not doing everything according to standards, so please be forgiving :)

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

https://github.com/ansible-collections/community.general/pull/1927 introduced deprecation for the `show_diff` parameter on the puppet module, it seems due to it being undocumented.

Since I believe it's worthwhile for a user to have the possibility to get change details when running puppet (not by default obviously, due to potential security concerns), I'm trying here to fix the documentation for that parameter to then be able to remove the deprecation.

Worth noting I haven't been able to run the sanity check yet (as described in the quick start guide):
```
[...]ansible_collections/community/general (puppet/show_diff) $ ansible-test sanity plugins/modules/system/puppet.py --docker -v
Traceback (most recent call last):
  File "~/.local/bin/ansible-test", line 69, in <module>
    main(context)
  File "~/.local/bin/ansible-test", line 38, in main
    generate_dockerfile(context)
  File "~/.local/bin/ansible-test", line 17, in generate_dockerfile
    resource_string("ansible_test","resources/Dockerfile.j2")
  File "~/.local/lib/python3.6/site-packages/jinja2/environment.py", line 1195, in __new__
    return env.from_string(source, template_class=cls)
  File "~/.local/lib/python3.6/site-packages/jinja2/environment.py", line 1092, in from_string
    return cls.from_code(self, self.compile(source), gs, None)
  File "~/.local/lib/python3.6/site-packages/jinja2/environment.py", line 750, in compile
    source = self._generate(source, name, filename, defer_init=defer_init)
  File "~/.local/lib/python3.6/site-packages/jinja2/environment.py", line 684, in _generate
    optimized=self.optimized,
  File "~/.local/lib/python3.6/site-packages/jinja2/compiler.py", line 112, in generate
    raise TypeError("Can't compile non template nodes")
TypeError: Can't compile non template nodes
```

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
Puppet

felixfontein commented:
Thanks for your contribution! Could you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)?

I'm interested in what the actual output is when setting `show_diff` to `true` (compared to `show_diff=false`). Can you give an example?

felixfontein commented on changelogs/fragments/3980-puppet-show_diff.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - puppet - add documentation for parameter ``show_diff`` and remove deprecation warning (https://github.com/ansible-collections/community.general/pull/3980).
```
```suggestion
  - puppet - remove deprecation for ``show_diff`` parameter (https://github.com/ansible-collections/community.general/pull/3980).
```

felixfontein commented on plugins/modules/system/puppet.py:
```diff
@@ -174,9 +179,7 @@ def main():
-            show_diff=dict(
-                type='bool', default=False, aliases=['show-diff'],
-                removed_in_version='7.0.0', removed_from_collection='community.general'),
+            show_diff=dict(type='bool', default=False, aliases=['show-diff']),
```
We should still deprecate that alias:
```suggestion
            show_diff=dict(
                type='bool', default=False, aliases=['show-diff'],
                deprecated_aliases=[dict(name='show-diff', version='7.0.0', collection_name='community.general')])),
```

vaudelbenoit replied:
This introduced a syntax error I fixed through b0f749d9

felixfontein commented on plugins/modules/system/puppet.py:
```diff
@@ -92,6 +92,11 @@
+    description:
+      - Wether to print file changes details
+    type: bool
+    default: false
```
```suggestion
    default: false
    aliases: [show-diff]
```

vaudelbenoit commented:
> Thanks for your contribution! ~Could you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)?~
> 
> I'm interested in what the actual output is when setting `show_diff` to `true` (compared to `show_diff=false`). Can you give an example?

Sure.

Ansible code:
```
- name: Run puppet
  puppet:
    timeout: "120m"
    noop: "{{ puppet_noop }}"
  register: puppet_result
  when: puppet_logdest is undefined
  failed_when: puppet_result.failed and not (puppet_ignore_errors_thisrun|bool
```

Without `show_diff`:
```
TASK [puppetrun : Run puppet] *************************************************************************************************************************************************************************************
Thursday 06 January 2022  01:46:06 +1300 (0:00:00.046)       0:00:22.591 ****** 
changed: [node]
node: stdout: Notice: Applied catalog in 25.02 seconds
Info: Using environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Info: Caching catalog for node
Info: Applying configuration version '1641386779'
Info: Computing checksum on file /etc/apt/apt.conf.d/15update-stamp
Info: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]: Filebucketed /etc/apt/apt.conf.d/15update-stamp to puppet with sum b9de0ac9e2c9854b1bb213e362dc4e41
Notice: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]/content: content changed '{md5}b9de0ac9e2c9854b1bb213e362dc4e41' to '{md5}0962d70c4ec78bbfa6f3544ae0c41974' (corrective)
Info: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]: Scheduling refresh of Class[Apt::Update]
Info: Class[Apt::Update]: Scheduling refresh of Exec[apt_update]
Notice: /Stage[main]/Apt::Update/Exec[apt_update]: Triggered 'refresh' from 1 event
Notice: Applied catalog in 25.02 seconds
```

With `show_diff: true`:
```
TASK [puppetrun : Run puppet] *************************************************************************************************************************************************************************************
Thursday 06 January 2022  01:48:38 +1300 (0:00:00.046)       0:00:25.365 ****** 
[DEPRECATION WARNING]: Param 'show_diff' is deprecated. See the module docs for more information. This feature will be removed from community.general in version 7.0.0. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
changed: [node]
node: stdout: Notice: Applied catalog in 11.91 seconds
Info: Using environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Info: Caching catalog for node
Info: Applying configuration version '1641386932'
Notice: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]/content: 
--- /etc/apt/apt.conf.d/15update-stamp	2022-01-06 01:47:00.477248729 +1300
+++ /tmp/puppet-file20220106-5003-19dz66e	2022-01-06 01:48:58.859055033 +1300
@@ -1 +1,2 @@
+// This file is managed by Puppet. DO NOT EDIT.
 APT::Update::Post-Invoke-Success {"touch /var/lib/apt/periodic/update-success-stamp 2>/dev/null || true";};
Info: Computing checksum on file /etc/apt/apt.conf.d/15update-stamp
Info: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]: Filebucketed /etc/apt/apt.conf.d/15update-stamp to puppet with sum b9de0ac9e2c9854b1bb213e362dc4e41
Notice: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]/content: 
Notice: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]/content: content changed '{md5}b9de0ac9e2c9854b1bb213e362dc4e41' to '{md5}0962d70c4ec78bbfa6f3544ae0c41974' (corrective)
Info: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]: Scheduling refresh of Class[Apt::Update]
Info: Class[Apt::Update]: Scheduling refresh of Exec[apt_update]
Notice: /Stage[main]/Apt::Update/Exec[apt_update]: Triggered 'refresh' from 1 event
Notice: Applied catalog in 11.91 seconds
```

With `show_diff: false`:
```
TASK [puppetrun : Run puppet] *************************************************************************************************************************************************************************************
Thursday 06 January 2022  01:50:23 +1300 (0:00:00.051)       0:00:25.971 ****** 
[DEPRECATION WARNING]: Param 'show_diff' is deprecated. See the module docs for more information. This feature will be removed from community.general in version 7.0.0. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
changed: [node]
node: stdout: Notice: Applied catalog in 12.04 seconds
Info: Using environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Info: Caching catalog for node
Info: Applying configuration version '1641387037'
Info: Computing checksum on file /etc/apt/apt.conf.d/15update-stamp
Info: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]: Filebucketed /etc/apt/apt.conf.d/15update-stamp to puppet with sum b9de0ac9e2c9854b1bb213e362dc4e41
Notice: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]/content: content changed '{md5}b9de0ac9e2c9854b1bb213e362dc4e41' to '{md5}0962d70c4ec78bbfa6f3544ae0c41974' (corrective)
Info: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]: Scheduling refresh of Class[Apt::Update]
Info: Class[Apt::Update]: Scheduling refresh of Exec[apt_update]
Notice: /Stage[main]/Apt::Update/Exec[apt_update]: Triggered 'refresh' from 1 event
Notice: Applied catalog in 12.04 seconds
```

vaudelbenoit commented:
I'm down to that python identifier with an hyphen in argument_spec. I can find other modules in the same case (composer has several of those for example), so surely this means it's more of a `warning` than an `error`, right?

felixfontein commented on tests/sanity/ignore-2.10.txt:
```diff
@@ -48,8 +48,6 @@ plugins/modules/system/iptables_state.py validate-modules:undocumented-parameter
 plugins/modules/system/osx_defaults.py validate-modules:parameter-state-invalid-choice
 plugins/modules/system/parted.py validate-modules:parameter-state-invalid-choice
 plugins/modules/system/puppet.py use-argspec-type-path
```
You need to add
```suggestion
plugins/modules/system/puppet.py validate-modules:parameter-invalid  # invalid alias - removed in 7.0.0
```
here and in the other ignore.txt files. (See here why it doesn't complain about composer: https://github.com/ansible-collections/community.general/blob/053b8a064db08b3c428b37a86b630f8e7b62b15d/tests/sanity/ignore-2.10.txt#L26)

vaudelbenoit replied:
Ha right, ty!

vaudelbenoit replied:
I put back `use-argspec-type-path` through another commit.

felixfontein commented on changelogs/fragments/3980-puppet-show_diff.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - puppet - remove deprecation for ``show_diff`` parameter (https://github.com/ansible-collections/community.general/pull/3980).
```
I think we should mention that the alias is still deprecated:
```suggestion
  - puppet - remove deprecation for ``show_diff`` parameter. Its alias ``show-diff`` is still deprecated and will be removed in community.general 7.0.0 (https://github.com/ansible-collections/community.general/pull/3980).
```
(This doesn't need a new `deprecated_features` entry IMO since it was already deprecated before, as part of the whole deprecated option.)

vaudelbenoit replied:
Sure, seems clearer.

felixfontein commented on plugins/modules/system/puppet.py:
```diff
@@ -176,7 +183,7 @@ def main():
             logdest=dict(type='str', default='stdout', choices=['all', 'stdout', 'syslog']),
             show_diff=dict(
```
Would you mind adding this comment? This will help the next person wondering what exactly this does :)
```suggestion
            # The following is not related to Ansible's diff; see https://github.com/ansible-collections/community.general/pull/3980#issuecomment-1005666154
            show_diff=dict(
```

vaudelbenoit commented:
Now that all checks are passing and I included the changes I was interested in, I'll leave the WIP tag on this PR for a couple of days in case someone else wants to comment, then I'll get rid of it (unless I'm told otherwise).

felixfontein commented:
@vaudelbenoit there's no need to keep the WIP there; usually WIP keeps people from looking at PRs since they expect more changes :)

vaudelbenoit commented:
> @vaudelbenoit there's no need to keep the WIP there; usually WIP keeps people from looking at PRs since they expect more changes :)

Ack'd, I'm removing it now then. Cheers!

felixfontein commented:
I'll merge this on Monday if nobody has further comments until then.

felixfontein commented:
@vaudelbenoit thanks for your contribution!
